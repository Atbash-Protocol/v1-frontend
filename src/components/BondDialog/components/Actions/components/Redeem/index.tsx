import { Box } from '@mui/material';

export const Redeem = ({ bondID }: { bondID: string }) => {
    return <Box>Redeem {bondID}</Box>;
};
