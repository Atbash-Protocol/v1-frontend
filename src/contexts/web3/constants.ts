export const USER_REJECTED = "User Rejected";
