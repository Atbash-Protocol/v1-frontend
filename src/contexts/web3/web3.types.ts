import { providers } from 'ethers';
import Web3Modal from 'web3modal';

export interface WEB3State {
    networkID: number | null;
    provider: providers.JsonRpcProvider | null;
    signer: providers.Web3Provider | null;
    signerAddress: string | null;
    web3Modal: Web3Modal;
}

export interface IWeb3Context {
    state: WEB3State;
    memoConnect: () => void;
    memoDisconnect: (currentSigner: WEB3State['signer']) => Promise<void>;
}

export enum WEB3ActionTypesEnum {
    CLOSE = 'CLOSE',
    NETWORK_CHANGED = 'NETWORK_CHANGED',
    SET_SIGNER = 'SET_SIGNER',
    CLEAR_SIGNER = 'CLEAR_SIGNER',
    SET_PROVIDER = 'SET_PROVIDER',
}

export interface WEB3ContextAction {
    type: WEB3ActionTypesEnum;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    payload?: any; // TODO : better typing
}
