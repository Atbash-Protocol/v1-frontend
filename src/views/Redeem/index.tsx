import { Typography } from '@mui/material';
import { useTranslation } from 'react-i18next';

import { BCard } from 'components/BCard';

function Redeem() {
    const { t } = useTranslation();

    return (
        <BCard title={t('bond:Redeem')} zoom={true}>
            <Typography> {t('ComingSoon')} </Typography>
        </BCard>
    );
}

export default Redeem;
