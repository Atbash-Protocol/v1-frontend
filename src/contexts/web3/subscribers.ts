import { Dispatch } from 'react';

import { providers } from 'ethers';
import Core from 'web3modal';

import { DEFAULT_NETWORK } from 'constants/blockchain';
import { getProviderURI } from 'contexts/web3/web3.utils';

import { WEB3ContextAction, WEB3ActionTypesEnum } from './web3.types';

export const subscribeSigner = async (web3provider: providers.Web3Provider, dispatch: Dispatch<WEB3ContextAction>) => {
    web3provider.on('networkChanged', async (networkId: number) => {
        console.log('sub', networkId);
        const signer = new providers.Web3Provider(web3provider as unknown as providers.ExternalProvider);

        dispatch({ type: WEB3ActionTypesEnum.NETWORK_CHANGED, payload: { signer, networkId } });
    });

    web3provider.once('close', () => {
        dispatch({ type: WEB3ActionTypesEnum.CLOSE, payload: { signer: web3provider.getSigner() } });
    });
};

export const createSigner = async (web3Modal: Core, dispatch: Dispatch<WEB3ContextAction>) => {
    const web3Provider = await web3Modal.connect();

    const signer = new providers.Web3Provider(web3Provider);

    subscribeSigner(web3Provider, dispatch);

    try {
        const [{ chainId }, address] = await Promise.all([signer.getNetwork(), signer.getSigner().getAddress()]);

        dispatch({ type: WEB3ActionTypesEnum.SET_SIGNER, payload: { signer, chainId, address } });
    } catch (err) {
        console.error(err);
    } finally {
    }
};

export const subscribeProvider = async (dispatch: Dispatch<WEB3ContextAction>) => {
    const provider = new providers.JsonRpcProvider(getProviderURI(DEFAULT_NETWORK));
    provider.on('error', err => {
        alert('provider error');
        console.error(err);
    });

    const { chainId } = await provider.ready;

    if (!chainId) throw new Error('Impossible to connect to the given provider');

    dispatch({ type: WEB3ActionTypesEnum.SET_PROVIDER, payload: { chainId, provider } });
};

export const resetWeb3Signer = (dispatch: Dispatch<WEB3ContextAction>, web3Modal: Core): void => {
    web3Modal.clearCachedProvider();
    localStorage.setItem('WEB3_CONNECT_CACHED_PROVIDER', '');
    dispatch({ type: WEB3ActionTypesEnum.CLOSE, payload: null });
};
