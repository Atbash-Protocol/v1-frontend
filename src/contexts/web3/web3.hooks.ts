import { useContext, useEffect, useMemo, useState } from 'react';

import { Dispatch } from 'redux';

import { DEFAULT_NETWORK } from 'constants/blockchain';
import { Web3Context } from 'contexts/web3/web3.context';
import { walletConnectWarning } from 'store/modules/messages/messages.slice';

export const useGoodNetworkCheck = () => {
    const {
        state: { networkID },
    } = useContext(Web3Context);

    const [isNetworkOk, setIsNetworkOk] = useState(false);

    useEffect(() => {
        if (networkID) setIsNetworkOk(networkID === DEFAULT_NETWORK);
    }, [networkID]);

    return isNetworkOk;
};

export const useSignerConnected = () => {
    const {
        state: { signer, networkID },
    } = useContext(Web3Context);

    const [isConnected, setIsConnected] = useState(signer !== null);

    useEffect(() => {
        if (signer && networkID) {
            setIsConnected(true);
        }
    }, [isConnected, signer, networkID]);

    return isConnected;
};

export const useWeb3ContextInitialized = () => {
    const {
        state: { signer, networkID, provider },
    } = useContext(Web3Context);

    const [isContextInitialized, setIsContextInitialized] = useState([provider, signer, networkID].some(e => e !== null));

    useEffect(() => {
        if (provider || signer) {
            setIsContextInitialized(true);
        }
    }, [provider, signer]);

    return isContextInitialized;
};

export const useSignerAddress = () => {
    const {
        state: { signerAddress },
    } = useContext(Web3Context);

    const [signer, setSigner] = useState(signerAddress);

    useEffect(() => {
        if (signerAddress) {
            setSigner(signerAddress);
        }
    }, [signerAddress]);

    return signer;
};

export const useProvider = () => {
    const {
        state: { provider },
    } = useContext(Web3Context);

    return provider;
};

export const useSafeSigner = (dispatch?: Dispatch) => {
    const {
        state: { signer, signerAddress },
    } = useContext(Web3Context);

    const memoSigner = useMemo(() => {
        if (!signer || !signerAddress) {
            if (dispatch) dispatch(walletConnectWarning);

            throw new Error('Missing signer or signerAddress ');
        }

        return { signer, signerAddress };
    }, [signer, signerAddress]);

    return memoSigner;
};
