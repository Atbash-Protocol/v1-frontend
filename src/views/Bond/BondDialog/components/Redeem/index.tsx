import { useEffect } from 'react';

import { Box, Button, Divider } from '@mui/material';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';

import Loader from 'components/Loader';
import MemoInlineMetric from 'components/Metrics/InlineMetric';
import { theme } from 'constants/theme';
import { useWeb3Context } from 'contexts/web3/web3.context';
import { selectBondsReady, selectBondItemMetrics, selectBondRedeemMetrics, selectBondQuoteResult } from 'store/modules/bonds/bonds.selector';
import { calcBondDetails, redeemBond } from 'store/modules/bonds/bonds.thunks';
import { BondMetrics } from 'store/modules/bonds/bonds.types';
import { selectIsRedeeming, selectIsRedeemingAndAutoStaking } from 'store/modules/transactions/transactions.selectors';
import { RootState } from 'store/store';

export const Redeem = ({ bondID, metrics, recipientAddress }: { bondID: string; metrics: BondMetrics; recipientAddress: string }) => {
    const dispatch = useDispatch();
    const { t } = useTranslation();
    const { bondDiscount, vestingTerm } = selectBondRedeemMetrics(metrics);

    const bondQuote = useSelector((state: RootState) => selectBondQuoteResult(state, t));
    const isClaiming = useSelector(selectIsRedeeming);
    const isClaimingAndStaking = useSelector(selectIsRedeemingAndAutoStaking);

    const {
        state: { networkID, signer },
    } = useWeb3Context();

    const bondMetrics = [
        { value: bondQuote.interestDue, metricKey: t('bond:PendingRewards') },
        { value: bondQuote.pendingPayout, metricKey: t('bond:PendingPayout') },
        { value: bondQuote.vesting, metricKey: t('bond:TimeFullyVested') },
        { value: bondDiscount, metricKey: t('ROI') },
        { value: t('common:day', { count: vestingTerm || 0 }), metricKey: t('bond:VestingTerm') },
    ].map(({ value, metricKey }, index) => <MemoInlineMetric {...{ value, metricKey }} key={index} />);

    useEffect(() => {
        if (networkID) {
            dispatch(calcBondDetails({ bondID, value: 0, networkID }));
        }
    }, [networkID]);

    const handleClaim = ({ isAutoStake }: { isAutoStake: boolean }) => {
        if (signer) {
            dispatch(
                redeemBond({
                    recipientAddress,
                    bondID,
                    isAutoStake,
                    signer,
                }),
            );
        }
    };

    return (
        <Box>
            <Box sx={{ justifyContent: 'center', display: 'flex', alignItems: 'center', flexDirection: 'column', mb: theme.spacing(2) }}>
                <Box mt={theme.spacing(2)} width={'50%'} p={theme.spacing(2)}>
                    <Button sx={{ width: '5rem' }} onClick={() => handleClaim({ isAutoStake: false })}>
                        {isClaiming ? t('common:PendingEllipsis') : t('bond:Claim')}
                    </Button>
                </Box>

                <Box mt={theme.spacing(2)} width={'50%'} p={theme.spacing(2)}>
                    <Button sx={{ width: '5rem' }} onClick={() => handleClaim({ isAutoStake: true })}>
                        {isClaimingAndStaking ? t('common:PendingEllipsis') : t('bond:ClaimAutostake')}
                    </Button>
                </Box>
            </Box>

            <Divider variant="fullWidth" textAlign="center" sx={{ borderColor: theme.palette.primary.light, marginBottom: theme.spacing(2) }} />
            <Box>{bondMetrics}</Box>
        </Box>
    );
};

const RedeemLoader = ({ bondID, recipientAddress }: { bondID: string; recipientAddress: string }) => {
    const bondsReady = useSelector(selectBondsReady);
    const metrics = useSelector((state: RootState) => selectBondItemMetrics(state, bondID));

    if (!bondsReady) return <Loader />;
    return <Redeem bondID={bondID} metrics={metrics} recipientAddress={recipientAddress} />;
};

export default RedeemLoader;
