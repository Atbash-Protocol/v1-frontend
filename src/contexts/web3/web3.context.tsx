import { createContext, useCallback, useContext, useEffect, useReducer } from 'react';

import { initWeb3Modal } from 'contexts/web3/web3.utils';

import { USER_REJECTED } from './constants';
import { subscribeProvider, createSigner, resetWeb3Signer } from './subscribers';
import { WEB3ActionTypesEnum, Web3Context, WEB3ContextAction, WEB3State } from './web3.types';

export const initialState: WEB3State = {
    provider: null,
    signer: null,
    signerAddress: null,
    networkID: null,
    web3Modal: initWeb3Modal(),
};

export const PWeb3Context = createContext<Web3Context>({
    state: initialState,
    memoConnect: Promise.reject,
    memoDisconnect: Promise.reject,
});

export const WEB3Reducer = (state: WEB3State, action: WEB3ContextAction) => {
    switch (action.type) {
        case WEB3ActionTypesEnum.CLOSE:
            return { ...state, signer: null, web3Modal: initWeb3Modal() };
        case WEB3ActionTypesEnum.NETWORK_CHANGED:
            return { ...state, networkID: Number(action.payload) };
        case WEB3ActionTypesEnum.SET_SIGNER:
            return { ...state, signer: action.payload.signer, signerAddress: action.payload.address, networkID: action.payload.chainId };
        case WEB3ActionTypesEnum.SET_PROVIDER:
            return { ...state, networkID: action.payload.chainId, provider: action.payload.provider };
        default:
            return state;
    }
};

export const NewWeb3ContextProvider = ({ children }: { children: JSX.Element }) => {
    const {
        state: { web3Modal, signer, provider },
    } = useContext(PWeb3Context);
    const [state, dispatch] = useReducer(WEB3Reducer, initialState);

    const memoConnect = useCallback(async () => {
        if (web3Modal && !signer) {
            await createSigner(web3Modal, dispatch);
        }
    }, [signer, web3Modal]);

    const memoDisconnect = useCallback(
        async (currentSigner: WEB3State['signer']) => {
            if (currentSigner && web3Modal) await resetWeb3Signer(dispatch, web3Modal);
        },
        [web3Modal],
    );

    useEffect(() => {
        // happens when user reload when metamask popup appears
        if (web3Modal) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            web3Modal.on('error', (err: any) => {
                if (err && err?.message && err?.message === USER_REJECTED) alert('User rejected connection');
            });
        }

        // forces the connection
        if (web3Modal && web3Modal.cachedProvider?.length > 0) {
            memoConnect();
        } else {
            if (!provider) {
                subscribeProvider(dispatch); // init at the provider
            }
        }
    }, [web3Modal]);

    // useEffect(() => {
    //     if (!signer && !provider) {
    //         console.log("Reaching this context");
    //         subscribeProvider(dispatch); // init at the provider
    //     }
    // }, [provider, signer]);

    return (
        <PWeb3Context.Provider
            value={{
                state,
                memoConnect,
                memoDisconnect,
            }}
        >
            {children}
        </PWeb3Context.Provider>
    );
};

export const usePWeb3Context = () => {
    return useContext(PWeb3Context);
};
