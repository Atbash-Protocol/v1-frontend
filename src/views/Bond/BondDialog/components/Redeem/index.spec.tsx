import { fireEvent, render, screen } from '@testing-library/react';
import { DateTime } from 'luxon';
import * as ReactReduxModule from 'react-redux';
import createMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { Web3Context } from 'contexts/web3/web3.context';
import { ContractEnum } from 'store/modules/app/app.types';

import Redeem from '.';

jest.mock('../BondPurchase', () => () => <> Bond Purchase</>);

const renderComponent = (component: JSX.Element, state: any, contextState?: any) => {
    const middlewares = [thunk];
    const mockStore = createMockStore(middlewares);
    const store = mockStore(state);

    const container = render(
        <Web3Context.Provider value={contextState}>
            <ReactReduxModule.Provider store={store}>{component}</ReactReduxModule.Provider>)
        </Web3Context.Provider>,
    );

    return { container, store };
};

describe('Redeem', () => {
    describe('when the state is not ready', () => {
        it('returns the loader', () => {
            const state = {
                bonds: {
                    bondMetrics: {},
                    bondInstances: {},
                },
                main: {
                    contracts: {},
                },
            };
            const { container } = renderComponent(<Redeem bondID="dai" recipientAddress="0x" />, state);

            expect(container.container.getElementsByClassName('MuiCircularProgress-root')).toHaveLength(1);
        });
    });

    describe('When the state is ready', () => {
        it('returns the Redeem component', () => {
            const state = {
                bonds: {
                    bondMetrics: {
                        dai: {},
                    },
                    bondInstances: {},
                    bondQuote: {
                        interestDue: 0,
                    },
                },
                main: {
                    blockchain: {
                        timestamp: DateTime.utc().toMillis(),
                    },
                    contracts: {
                        [ContractEnum.BASH_CONTRACT]: jest.fn(),
                    },
                },
                transactions: [],
            };

            const contextState = { state: { networkID: 1, signer: jest.fn() } };

            const { container, store } = renderComponent(<Redeem bondID="dai" recipientAddress="0x" />, state, contextState);

            expect(container.container).toMatchSnapshot();
            expect(store.getActions()).toHaveLength(1);

            const btns = screen.getAllByText(/bond:Claim/i);
            expect(btns).toHaveLength(2);

            fireEvent.click(btns[0]);

            expect(store.getActions().at(1)).toEqual({
                type: 'bonds/redeem/pending',
                payload: undefined,
                meta: {
                    arg: {
                        bondID: 'dai',
                        isAutoStake: false,
                        recipientAddress: '0x',
                        signer: expect.anything(),
                    },
                    requestId: expect.any(String),
                    requestStatus: 'pending',
                },
            });

            fireEvent.click(btns[1]);
            expect(store.getActions()).toHaveLength(3);
            expect(store.getActions().at(2)).toEqual({
                type: 'bonds/redeem/pending',
                payload: undefined,
                meta: {
                    arg: {
                        bondID: 'dai',

                        isAutoStake: true,
                        recipientAddress: '0x',
                        signer: expect.anything(),
                    },
                    requestId: expect.any(String),
                    requestStatus: 'pending',
                },
            });
        });
    });
});
