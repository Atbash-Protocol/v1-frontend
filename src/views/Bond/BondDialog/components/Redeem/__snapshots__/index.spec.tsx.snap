// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redeem When the state is ready returns the Redeem component 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-1q2toq6"
    >
      <div
        class="MuiBox-root css-1blr7m7"
      >
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1lxq5oa-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          bond:Claim
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root css-1blr7m7"
      >
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1lxq5oa-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          bond:ClaimAutostake
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-fimgk4-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1z137pa"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
        >
          bond:PendingRewards
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
        >
          0.00 BASH
        </p>
      </div>
      <div
        class="MuiBox-root css-1z137pa"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
        >
          bond:PendingPayout
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-15tfpcm-MuiSkeleton-root"
          />
        </p>
      </div>
      <div
        class="MuiBox-root css-1z137pa"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
        >
          bond:TimeFullyVested
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-15tfpcm-MuiSkeleton-root"
          />
        </p>
      </div>
      <div
        class="MuiBox-root css-1z137pa"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
        >
          ROI
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-15tfpcm-MuiSkeleton-root"
          />
        </p>
      </div>
      <div
        class="MuiBox-root css-1z137pa"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
        >
          bond:VestingTerm
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
        >
          common:day
        </p>
      </div>
    </div>
  </div>
  )
</div>
`;
